### PR TITLE
Fix MonoSortKey native/managed layout mismatch.

### PR DESCRIPTION
**Pull request by jaykrell**
MonoSortKey native:
https://github.com/mono/mono/blob/f019810b12ba130b7fed1cd41bb2eaed04ececef/mono/metadata/object-internals.h#L700

SortKey managed:
https://github.com/mono/mono/blob/f019810b12ba130b7fed1cd41bb2eaed04ececef/mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs#L69

do not agree.

Memory is corrupted and crash if you set $MONO_DISABLE_MANAGED_COLLATION=yes documented here:
https://github.com/mono/mono/blob/f019810b12ba130b7fed1cd41bb2eaed04ececef/man/mono.1#L1161

Fix is to remove the native code and replace by trivial managed.

Found testing coop-conversion of System.Globalization. https://github.com/mono/mono/pull/12023

### CODE REVIEW

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mono/metadata/locales.c [hunk 1]
```diff
@@ -740,7 +740,8 @@ int ves_icall_System_Globalization_CompareInfo_internal_compare (MonoCompareInfo
 					 options));
 }
 
-void ves_icall_System_Globalization_CompareInfo_assign_sortkey (MonoCompareInfo *this_obj, MonoSortKey *key, MonoString *source, gint32 options)
+MonoArray*
+ves_icall_System_Globalization_CompareInfo_assign_sortkey (MonoString *source)
 {
 	ERROR_DECL (error);
```
Comment (vargaz) [inline mono/metadata/locales.c]: This doesn't seem to do anything useful anymore, it converts a string to its bytes.

Comment (jaykrell) [reply mono/metadata/locales.c]: Yeah you are right -- just an identity. Just rewrite it in all managed?

There are a few confusing things.
In particular, I thought I saw a similar crash in an unrelated build.
Like maybe monolite got updated with the change, and maybe we depend on the data being corrupted?
Maybe pretend I didn't see anything confusing and proceed.

Removing the icall and just doing it in managed seems reasonable. Let me try that.

Comment (jaykrell) [reply mono/metadata/locales.c]: Identity, but truncating guint16 to guint8, granted.

Comment (jaykrell) [reply mono/metadata/locales.c]: Ugh.
```
ReferenceSources/CompareInfo.cs(85,8): error CS0122: 'SortKey.key' is inaccessible due to its protection level
make[8]: *** [../../class/lib/build-macos/mscorlib.dll] Error 1
```

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs [hunk 2]
```diff
@@ -89,6 +98,7 @@ internal SortKey (int lcid, string source, CompareOptions opt)
 			this.source = source;
 			this.key = buffer;
 			this.options = opt;
+			this.key = CreateKey (source);
```
Comment (jaykrell) [inline mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs]: oops

**Comment by jaykrell** [comment]
@monojenkins build deb with monolite

File: mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs [hunk 3]
```diff
@@ -77,6 +77,11 @@ internal SortKey (int lcid, string source, CompareOptions opt)
 			this.lcid = lcid;
 			this.source = source;
 			this.options = opt;
+			var source_length = source.Length;
+			var k = new byte [source_length];
+			for (int i = 0; i < source_length; ++i)
+				k [i] = (byte)source [i];
```
Comment (lewurm) [inline mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs]: shorter?
```csharp
this.key = new byte [source.Length];
for (int i = 0; i < source.Length; ++i)
  this.key [i] = (byte) source [i];
```

Comment (jaykrell) [reply mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs]: I don't trust every `.` to be a reload from memory. Do you?
I agree it works and means the same thing of course.

**Comment by jaykrell** [comment]
corlib bump not clearly needed.
The new runtime does not work with old corlib. But could be made to, by leaving the code in.
New corlib does work with old runtime.

**Comment by jaykrell** [comment]
@monojenkins squash

File: mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs (hunk 3 above)
Comment (lewurm) [reply mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs]: is it critical in terms of perf or raciness?

Comment (jaykrell) [reply mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs]: No but I prefer to keep it faster when easy.

**Comment by jaykrell** [comment]
@monojenkins build Linux x64 Coop Suspend

**Comment by vargaz** [comment]
@monojenkins build failed

File: mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs (hunk 3 above)
Comment (luhenry) [reply mcs/class/corlib/Mono.Globalization.Unicode/SortKey.cs]: Use the shorter version.

**Pull request merged**